pkg/tests/api/alerting/api_notification_channel_test.go

```diff
@@ -0,0 +1,963 @@
+
+		// Post the alertmanager config.
+		u := fmt.Sprintf("http://grafana:password@%s/api/alertmanager/grafana/config/api/v1/alerts", grafanaListedAddr)
+		postRequest(t, u, amConfig, http.StatusAccepted)
+
+		// Verifying that all the receivers and routes have been registered.
+		alertsURL := fmt.Sprintf("http://grafana:password@%s/api/alertmanager/grafana/config/api/v1/alerts", grafanaListedAddr)
+		resp := getRequest(t, alertsURL, http.StatusOK) // nolint
+		b := getBody(t, resp.Body)
+		require.JSONEq(t, getExpAlertmanagerConfigFromAPI(mockChannel.server.Addr), b)
+	}
+
+	{
+		// Create rules that will fire as quickly as possible
+		u := fmt.Sprintf("http://grafana:password@%s/api/ruler/grafana/api/v1/rules/default", grafanaListedAddr)
+		postRequest(t, u, rulesConfig, http.StatusAccepted)
+	}
+
+	// Eventually, we'll get all the desired alerts.
+	// nolint:gosec
+	require.Eventually(t, func() bool {
+		return mockChannel.totalNotifications() == len(alertNames) &&
+			mockChannel.matchesExpNotifications(expNotifications)
+	}, 25*time.Second, 1*time.Second)
+
+	require.NoError(t, mockChannel.Close())
+}
+
+func getRequest(t *testing.T, url string, expStatusCode int) *http.Response {
+	// nolint:gosec
+	resp, err := http.Get(url)
+	t.Cleanup(func() {
+		require.NoError(t, resp.Body.Close())
+	})
+	require.NoError(t, err)
+	require.Equal(t, expStatusCode, resp.StatusCode)
+	return resp
+}
+
+func postRequest(t *testing.T, url string, body string, expStatusCode int) {
+	buf := bytes.NewReader([]byte(body))
+	// nolint:gosec
+	resp, err := http.Post(url, "application/json", buf)
+	t.Cleanup(func() {
+		require.NoError(t, resp.Body.Close())
+	})
+	require.NoError(t, err)
+	require.Equal(t, expStatusCode, resp.StatusCode)
+}
+
+type mockNotificationChannel struct {
+	t      *testing.T
+	server *http.Server
+
+	receivedNotifications    map[string][]string
+	receivedNotificationsMtx sync.Mutex
+}
+
+func newMockNotificationChannel(t *testing.T, grafanaListedAddr string) *mockNotificationChannel {
+	lastDigit := grafanaListedAddr[len(grafanaListedAddr)-1] - 48
+	lastDigit = (lastDigit + 1) % 10
+	newAddr := fmt.Sprintf("%s%01d", grafanaListedAddr[:len(grafanaListedAddr)-1], lastDigit)
+
+	nc := &mockNotificationChannel{
+		server: &http.Server{
+			Addr: newAddr,
+		},
+		receivedNotifications: make(map[string][]string),
+		t:                     t,
+	}
+
+	nc.server.Handler = nc
+	go func() {
+		require.Equal(t, http.ErrServerClosed, nc.server.ListenAndServe())
+	}()
+
+	return nc
+}
+
+func (nc *mockNotificationChannel) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	nc.receivedNotificationsMtx.Lock()
+	defer nc.receivedNotificationsMtx.Unlock()
+
+	urlParts := strings.Split(req.URL.String(), "/")
+	key := fmt.Sprintf("%s/%s", urlParts[len(urlParts)-2], urlParts[len(urlParts)-1])
+	body := getBody(nc.t, req.Body)
+
+	fmt.Println("--- $$", key, body)
+
+	// Special handling for non-static data, like timestamp.
```

Comment:
Sorry if this is a dumb question, but why do we have special handling in write path and then in the read path as well? I would expect us to handle special cases in one or the other (my preference is the read path as that way we preserve the original data)

Reply:
Moved it all to a single place - when checking expected to actual. And turns out I was missing 2 more special cases which I have fixed now.